tensor2tensor/data_generators/translate_enfr_src_feat.py

```diff
@@ -0,0 +1,178 @@
+based on any hparams defined for the transformer model.
+
+There can be an arbitrary number of source features. These are
+provided in a separate file, where words are separated by space
+and the word features by TranslateEnfrSrcFeat.sfeat_delimiter.
+
+Example with parts-of-speech and lemmas as source features ('|'
+as delimiter):
+source file: I hated cats .
+sfeats file: Pronoun|I Verb-Past|hate Noun-Plural|cat Punctuation|.
+
+Since all features must be synchronized at word level, the inputs
+(and outputs) must be already tokenized. Words (i.e. the first feature)
+are then split into subword units and source features are multiplied
+in order to keep the same unique sentence length. To add subword tags
+marking the beginning, inside and end of a word, set
+TranslateEnfrSrcFeat.use_subword_tags() below to True (and give a
+size to this last feature embedding in "source_feature_sizes").
+
+For inference, the file containing the features must be given in
+a separate argument for t2t-decoder:
+
+--source_feature_file=/path/to/my/source_feature_file
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+# Dependency imports
+
+from tensor2tensor.data_generators import problem
+from tensor2tensor.data_generators import translate_source_features
+from tensor2tensor.models import transformer
+from tensor2tensor.utils import registry
+
+import tensorflow as tf
+
+# Below is the small data for translate_enfr problem initially in:
+# https://s3.amazonaws.com/opennmt-trainingdata/baseline-1M-enfr.tgz
+# The data has been cleaned and tokenized using Moses scripts:
+# https://github.com/moses-smt/mosesdecoder/tree/master/scripts/training/clean-corpus-n.perl
+# https://github.com/moses-smt/mosesdecoder/tree/master/scripts/tokenizer/normalize-punctuation.perl
+# https://github.com/moses-smt/mosesdecoder/tree/master/scripts/tokenizer/tokenizer.perl
+# English parts-of-speech and lemmas were obtained with tree-tagger:
+# http://www.cis.uni-muenchen.de/~schmid/tools/TreeTagger/
+
+_ENFR_TRAIN_DATA = [
+    [
+        "https://dl.dropboxusercontent.com/s/d8i4qdrm01s5wef/data_onmt.tgz",
+        ("baseline-1M_train.clean.en.tok",
+         "baseline-1M_train.clean.fr.tok"),
+        "baseline-1M_train.clean.en.sft"
+    ],
+]
+_ENFR_DEV_DATA = [
+    [
+        "https://dl.dropboxusercontent.com/s/d8i4qdrm01s5wef/data_onmt.tgz",
+        ("baseline-1M_valid.clean.en.tok",
+         "baseline-1M_valid.clean.fr.tok"),
+        "baseline-1M_valid.clean.en.sft"
+    ],
+]
+
+
+@registry.register_hparams
+def transformer_sfeats_hparams():
+  # define initial transformer hparams here
+  hp = transformer.transformer_base()
+  #hp = transformer.transformer_big()
+
+  # feature vector size setting
+  # the order of the features is the same
+  # as in the source feature file. All
+  # sizes are separated by ':'
+  hp.add_hparam("source_feature_sizes", "16:56:8")
```

Comment:
@franckbrl I think an explanation of the feature sizes would be great here. I guess `56` is the size of available tags? I'm not sure, but when I use the following command:

```bash
cat baseline-1M_train.clean.en.sft | tr ' ' '\n' | cut -d "|" -f 1 | sort | uniq | wc -l
```

The reported size is `58` instead of `56`.

Reply:
`source_feature_size` sets the embedding size (number of units) of the feature. The vocabulary size is not set in advance, it is computed on the full training data. So here, the first feature embedding has a dimension 16, and the feature has a vocabulary of 58 PoS-tags.

Reply:
Ah, ok, I understand it now :) What do you think about renaming it to `source_feature_hidden_sizes`? Thanks for your patience!

Reply:
These are not `hidden` layers, since the embeddings given as input. We do have a specific hidden layer for the encoder, and its size is `enc_hidden_size`.

Reply:
Sorry, my mistake - I updated my previous comment. `source_feature_embedding_sizes` would be ok?

Reply:
This sounds good, and it can't be understood as vocabulary size anymore. Let's do it?

Reply:
Would be great :)

Reply:
Done!